egs/librispeech/asr/simple_v1/bpe_ctc_att_conformer_decode.py

```diff
@@ -6,83 +6,286 @@
+    # First, extract `num_paths` paths for each sequence.
+    # paths is a k2.RaggedInt with axes [seq][path][arc_pos]
+    paths = k2.random_paths(lats, num_paths=num_paths, use_double_scores=True)
+
+    # token_seqs/word_seqs is a k2.RaggedInt sharing the same shape as `paths`
+    # but it contains word IDs. Note that it also contains 0s and -1s.
+    # The last entry in each sublist is -1.
+    token_seqs = k2.index(lats.labels.contiguous(), paths)
+    word_seqs = k2.index(lats.aux_labels.contiguous(), paths)
+
+    # Note: the above operation supports also the case when
+    # lats.aux_labels is a ragged tensor. In that case,
+    # `remove_axis=True` is used inside the pybind11 binding code,
+    # so the resulting `word_seqs` still has 3 axes, like `paths`.
+    # The 3 axes are [seq][path][word]
+
+    # Remove epsilons and -1 from word_seqs
+    token_seqs = k2.ragged.remove_values_leq(token_seqs, 0)
+    word_seqs = k2.ragged.remove_values_leq(word_seqs, 0)
+
+    # Remove repeated sequences to avoid redundant computation later.
+    #
+    # Since k2.ragged.unique_sequences will reorder paths within a seq,
+    # `new2old` is a 1-D torch.Tensor mapping from the output path index
+    # to the input path index.
+    # new2old.numel() == unique_word_seqs.num_elements()
+    unique_token_seqs, _, new2old = k2.ragged.unique_sequences(
+        token_seqs, need_num_repeats=False, need_new2old_indexes=True)
+    # Note: unique_token_seqs still has the same axes as token_seqs
+
+    seq_to_path_shape = k2.ragged.get_layer(unique_token_seqs.shape(), 0)
+
+    # path_to_seq_map is a 1-D torch.Tensor.
+    # path_to_seq_map[i] is the seq to which the i-th path
+    # belongs.
+    path_to_seq_map = seq_to_path_shape.row_ids(1)
+
+    # Remove the seq axis.
+    # Now unique_word_seqs has only two axes [path][word]
+    unique_token_seqs = k2.ragged.remove_axis(unique_token_seqs, 0)
+
+    # token_fsas is an FsaVec with axes [path][state][arc]
+    token_fsas = k2.linear_fsa(unique_token_seqs)
+
+    token_fsas_with_epsilon_loops = k2.add_epsilon_self_loops(token_fsas)
+
+    # lats has token IDs as labels and word IDs as aux_labels.
+    # inv_lats has word IDs as labels and token IDs as aux_labels
+    # Do k2.invert to make it compatible to function compute_am_scores
+    inv_lats = k2.invert(lats)
+    inv_lats = k2.arc_sort(inv_lats) # no-op if inv_lats is already arc-sorted
+    am_scores = compute_am_scores(inv_lats, token_fsas_with_epsilon_loops, path_to_seq_map)
+
+    lats = k2.arc_sort(lats)
+    fgram_lm_lats = _intersect_device(lats, token_fsas_with_epsilon_loops, path_to_seq_map, sorted_match_a=True)
+    fgram_lm_lats = k2.top_sort(k2.connect(fgram_lm_lats.to('cpu')).to(lats.device))
+    # am_scores is computed with log_semiring=True
+    # set log_semiring=True here to make fgram_lm_scores comparable to am_scores
+    fgram_tot_scores = fgram_lm_lats.get_tot_scores(use_double_scores=True, log_semiring=True)
```

Comment:
Please see https://github.com/k2-fsa/snowfall/pull/214

> The 2nd arg to get_tot_scores() here, representing log_semiring, should be false, because ARPA-type language models are constructed in such a way that the backoff prob is included in the direct arc. I.e. we would be double-counting if we were to sum the probabilities of the non-backoff and backoff arcs.

Have you tried to use `log_semiring=False`?

Reply:
Not yet, will try it.

Reply:
log_semiring=False is a little better than log_semiring=True （3.84% vs. 3.86%) with num_paths=500.

INFO:root:[test-clean-lm_scale_0.6] %WER 2.84% [1491 / 52576, 203 ins, 135 del, 1153 sub ]

```
- fgram_tot_scores = fgram_lm_lats.get_tot_scores(use_double_scores=True, log_semiring=True)
+ fgram_tot_scores = fgram_lm_lats.get_tot_scores(use_double_scores=True, log_semiring=False)
```